### ISSUE TITLE
SIGTERM when processing versions on heroku

### ISSUE THREAD

**Issue by atstockland** [lardawge/carrierwave_backgrounder#255]
I have been chasing this error and am beginning to think it may be a bug.  I apologize if it's not---and, I certainly don't want to treat the issue tracker like stack overflow.

I have an Expense model that uses a receipt uploader.  I use carrierwave_backgrounder to handle processing async.  carrierwave_backgrounder results in a failed job with error "SIGTERM".  The jobs actually do process successfully, because the versions are created.  However, receipt_processing column is never updated to false since the job never triggers after_perform.  I've been investigating this for a couple weeks and am beginning to wonder if it is a carrierwave_backgrounder bug or a problem in my implementation.

Using Rails 5.0.1, carrierwave (0.11.2), carrierwave_backgrounder (0.4.2), delayed_job (4.1.2).

As you will see, I use a base uploader to DRY up my uploaders.

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -  

I have an Expense model that uses the ReceiptUploader.

    expense.rb

      mount_uploader :receipt, ReceiptUploader
      process_in_background :receipt if Rails.env.production?
      attr_accessor :receipt_cache

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 
I can include the full stack trace (pretty long) from the error if needed.

    Failed Job.

      ID: 2691222
      Priority: 3
      Attempts: 1
      Queue: carrierwave_backgrounder
      Handler: --- !ruby/struct:CarrierWave::Workers::ProcessAsset
        klass: Expense
        id: '701'
        column: :receipt
      Last Error: SIGTERM /app/vendor/bundle/ruby/2.3.0/gems/delayed_job-4.1.2/lib/delayed/worker.rb:160:in `block in s...
      Run At: 3 minutes ago
      Locked At:  3 minutes ago
      Locked By: host:a1148aa4-9539-4859-992c-8e4a66dfbe59 pid:4

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
      Uploader:

      class ReceiptUploader < CarrierWave::Uploader::Base
        include CarrierWave::MiniMagick
        include ::CarrierWave::Backgrounder::Delay
        include CarrierWave::ImageOptimizer

        include Uploadable # base uploader, see module below
        include SmallVersion # see this module below
        include LargeVersion # # see this module below

        def store_dir
          "users/#{model.user_id}/receipts/#{model.created_at.year}/#{model.expense_code}"
        end

        def filename
          extension = ".#{model.receipt.file.extension}" if model.receipt? && model.receipt.file.extension.present?
          "#{model.expense_code}_#{model.user_id}#{extension}".downcase if original_filename
        end

    end

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
      Version: 

      module SmallVersion
        extend ActiveSupport::Concern
        include Uploadable # base uploader
        include CarrierWave::MiniMagick

      included do

        version :small do
        process convert: 'png'
        process :set_png_content_type
        process optimize: [{ quality: 65, quiet: true }]
        process resize_to_limit: [10000, 200]
        def full_filename (for_file = model.receipt.file)
          super.chomp(File.extname(super)) + '.png'
        end
      end

      end
      end

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
      Base Uploader:

      module Uploadable
        extend ActiveSupport::Concern

       included do

         def extension_whitelist
          %w(pdf jpeg jpg tif tiff png gif)
         end

         def content_type_whitelist
           [/(image|application)\//]
         end

         def content_type_blacklist
           ['application/text', 'application/json']
         end

         private

         def set_jpg_content_type(*_args)
           file.instance_variable_set(:@content_type, 'image/jpg')
         end

         def set_gif_content_type(*_args)
           file.instance_variable_set(:@content_type, 'image/gif')
         end

         def set_png_content_type(*_args)
           file.instance_variable_set(:@content_type, 'image/png')
         end

         def image?(new_file)
           new_file.content_type.start_with? 'image'
         end

         def pdf?(new_file)
           new_file.content_type &&
           new_file.content_type == 'application/pdf'
         end
     end

     end

**Comment by atstockland** [comment]
i'm gonna close this---I think i have tracked down whats happening.  I found a couple errors in heroku logs that I need to address.

Jan 11 22:22:08 xxx heroku/worker.2:  Process running mem=800M(156.4%) 
Jan 11 22:22:08 xxx heroku/worker.2:  Error R14 (Memory quota exceeded) 
Jan 11 22:22:09 xxx app/worker.1:  [Worker(host:5d67f8aa-419f-4425-9a22-7cf90cf2818a pid:4)] Job CarrierWave::Workers::ProcessAsset (id=2691225) RUNNING 
Jan 11 22:22:10 xxx app/worker.1:  convert.im6: invalid profile length `/tmp/mini_magick20170112-4-3iqifa.jpg' @ warning/png.c/MagickPNGWarningHandler/1754. 
Jan 11 22:22:11 xxx app/worker.1:  Attempting to optimize a jpeg without jpegoptim installed. Skipping... 
Jan 11 22:22:12 xxx app/worker.1:  convert.im6: invalid profile length `/tmp/mini_magick20170112-4-1rp9yop.jpg' @ warning/png.c/MagickPNGWarningHandler/1754.

**Comment by atstockland** [comment]
I have reopened this.  After cleaning up other warnings the same issue persist.  It seems the process actions are using a lot of memory in the background worker.  Is this in any way an issue with CB?  I have found some similar issues regarding how heroku handles workers resulting in SIGTERM and SIGKILL.  My other background jobs run fine, only jobs from CB run slowly.  Like I said before, the jobs eventually complete (in that versions are processed), but the _processing flag is never reset to false.  Thanks!
